sale.py

```diff
@@ -226,3 +245,120 @@ def process(cls, sales):
             if sale.state not in ('confirmed', 'processing', 'done'):
                 continue        # pragma: no cover
             sale.create_gift_cards()
+
+
+class Payment:
+    'Payment'
+    __name__ = 'sale.payment'
+
+    gift_card = fields.Many2One(
+        "gift_card.gift_card", "Gift Card", states={
+            'required': Eval('method') == 'gift_card',
+            'invisible': ~(Eval('method') == 'gift_card'),
+        }, domain=[('state', '=', 'active')], depends=['method']
+    )
+
+    def _create_payment_transaction(self, amount, description):
+        """Creates an active record for gateway transaction.
+        """
+        PaymentTransaction = Pool().get('payment_gateway.transaction')
+        Date = Pool().get('ir.date')
+
+        return PaymentTransaction.create([{
+            'description': description or 'Auto charge from sale',
+            'date': Date.today(),
+            'party': self.sale.party,
+            'payment_profile': self.payment_profile,
+            'address': (
+                self.payment_profile and
+                self.payment_profile.address or self.sale.invoice_address),
+            'amount': self.sale.currency.round(amount),
+            'currency': self.sale.currency,
+            'gateway': self.gateway,
+            'sale_payment': self.id,
+            'provider_reference': self.reference,
+            'gift_card': self.gift_card and self.gift_card.id or None
+        }])[0]
+
+    @classmethod
+    def validate(cls, payments):
+        """
+        Validate sale payment to add gift card with sufficient amount only
+        """
+        super(Payment, cls).validate(payments)
+
+        for payment in payments:
+            if payment.gift_card:
+                payment.check_gift_card_amount()
+
+    def check_gift_card_amount(self):
```

Comment:
this validation would be triggered on all edits to `sale.payment` even after the payment is consumed. In the consumed state, the gift_card will never have a balance ? and this validation would always fail.

PaymentTransaction has the logic to check for gift card balances on authorization and capture and there is no reason why this should check here as well. OTOH, displaying available balance is an API choice.